### ISSUE TITLE
Web components should work on the server too

### ISSUE THREAD

**Issue by NilsJacobsen** [opral/inlang-message-ui-components#54]
We didn't need it for the settings component, but on the product pages on inlang.com this is a requirement. Right now I can only import the component when I am on the client. Otherwise it gives me a HTMLElement is not defined error. <br><br>Gonna try to make that possible.

**Comment by samuelstroschein** [comment]
You can use web components on the server with a lit's server side renderer [https://lit.dev/docs/ssr/overview/](https://lit.dev/docs/ssr/overview/).

I assume that the effort of two renderers (solid and lit) is not worth it. The web component will be rendered on the client with solid's renderer. But I think that's OK until inlang.com slowly/fully migrates to web components and replaces the solid renderer with lit's

**Comment by NilsJacobsen** [comment]
**TL;DR**

SSR is a requirement for for components. Otherwise it leads to flashing UI and worse SEO.<br>-> I can probably workaround the topic for this component by using slots.<br><br>Loom: [https://www.loom.com/share/bd7587447fa64ed19f4e92ed5d5a4efd](https://www.loom.com/share/bd7587447fa64ed19f4e92ed5d5a4efd)

**Comment by samuelstroschein** [comment]
I see three options to move forward: <br><br>Option 1 - Don't care for now (bad UX and SEO)<br><br>Option 2 - Your anchor tag hack (prob bad for SEO/mixing solid with lit) <br><br>Option 3 - Per page either solid or lit renderer (Higher effort, ensures that SEO will work)<br><br>**Option 3 - per page either solid or lit rendering<br>**<br>Use [https://lit.dev/docs/libraries/standalone-templates/](https://lit.dev/docs/libraries/standalone-templates/) to write the HTML and [https://www.npmjs.com/package/@lit-labs/ssr#rendering-to-a-string](https://www.npmjs.com/package/@lit-labs/ssr#rendering-to-a-string) to render it on the server. <br><br><br>Example

```
import { html, LitElement } from "lit"

export class Page extends LitElement {
  render() { 
    html`
    <div class="blablabla">
      <component-1></component-1>
    </div>
    `
}

@customElement('component-1')
class Component1 extends LitElement {

}
```

Pros:

* SSR etc will work because no mix of solid and lit
* Per page migration from Solid to Lit 

Cons: 

* more effort

**Comment by NilsJacobsen** [comment]
Gonna make a test to see if I can get it working on in our setup.

**Comment by samuelstroschein** [comment]
you can use an if statement 

```
if (page instaceof LitElement){
  litrenderer
} else {
  solidrenderer
}
```

**Comment by NilsJacobsen** [comment]
That was my plan 😀 I'm also gonna see how to hydrate it afterwards. Well I guess like that somehow [https://lit.dev/docs/ssr/client-usage/](https://lit.dev/docs/ssr/client-usage/)

**Comment by NilsJacobsen** [comment]
Well it's not that easy aka a bigger change, because the Page also uses the Navigation Header, Root, language stuff, which is not lit syntax.

**Comment by NilsJacobsen** [comment]
Btw option 2 is also ruled out because that would lead to a huge code duplication.

**Comment by samuelstroschein** [comment]
@nils.jacobsen i see. the layout would still mix solid and lit ergh :/

**Comment by samuelstroschein** [comment]
other option might be: migrate fully from solid to lit given that you don't have complex UI components to avoid the two-render approach?
